Integrate ltac2 (#10002)
coq/coq: doc/sphinx/proof-engine/ltac2.rst

.. _ltac2:

Ltac2
=====

The Ltac tactic language is probably one of the ingredients of the success of
Coq, yet it is at the same time its Achilles' heel. Indeed, Ltac:

- has often unclear semantics
- is very non-uniform due to organic growth
- lacks expressivity (data structures, combinators, types, ...)
- is slow
- is error-prone and fragile
- has an intricate implementation

Following the need of users that start developing huge projects relying
critically on Ltac, we believe that we should offer a proper modern language
that features at least the following:

- at least informal, predictable semantics
- a typing system
- standard programming facilities (i.e. datatypes)

> In this chapter, we describe the implementation of such a language. The previous

Reviewer: Fixed.

- standard programming facilities (i.e. datatypes)

This document describes the implementation of such a language. The
implementation of Ltac as of Coq 8.7 will be referred to as Ltac1.

.. _ltac2_design:

General design
--------------

There are various alternatives to Ltac1, such that Mtac or Rtac for instance.
While those alternatives can be quite distinct from Ltac1, we designed
Ltac2 to be closest as reasonably possible to Ltac1, while fixing the
aforementioned defects.

In particular, Ltac2 is:

- a member of the ML family of languages, i.e.

  * a call-by-value functional language
  * with effects
  * together with Hindley-Milner type system

- a language featuring meta-programming facilities for the manipulation of
  Coq-side terms
- a language featuring notation facilities to help writing palatable scripts

We describe more in details each point in the remainder of this document.

ML component
------------

Overview
~~~~~~~~

Ltac2 is a member of the ML family of languages, in the sense that it is an
effectful call-by-value functional language, with static typing à la

> Hindley-Milner. It is commonly accepted that ML constitutes a sweet spot in PL

Author: @all anybody has a reference for Hindley-Milner, so that I insert it?

> Hindley-Milner. It is commonly accepted that ML constitutes a sweet spot in PL

Reviewer 2: That's the common one https://dl.acm.org/citation.cfm?doid=582153.582176

> Hindley-Milner. It is commonly accepted that ML constitutes a sweet spot in PL

Author: Thanks @ejgallego!

design, as it is relatively expressive while not being either too lax
(contrarily to dynamic typing) nor too strict (contrarily to say, dependent
types).

The main goal of Ltac2 is to serve as a meta-language for Coq. As such, it
naturally fits in the ML lineage, just as the historical ML was designed as
the tactic language for the LCF prover. It can also be seen as a general-purpose
language, by simply forgetting about the Coq-specific features.

Sticking to a standard ML type system can be considered somewhat weak for a
meta-language designed to manipulate Coq terms. In particular, there is no
way to statically guarantee that a Coq term resulting from an Ltac2
computation will be well-typed. This is actually a design choice, motivated
by retro-compatibility with Ltac1. Instead, well-typedness is deferred to
dynamic checks, allowing many primitive functions to fail whenever they are

> provided with an ill-typed term.

Author: I think it would be easier to read a separate section about related work than inserting specific differences at each documentation point. I suggest we do that post-merge, though.

proof engine. This allows to think of proof-modifying primitives as effects
in a straightforward way. Semantically, proof manipulation lives in a monad,
which allows to ensure that Ltac2 satisfies the same equations as a generic ML
with unspecified effects would do, e.g. function reduction is substitution
by a value.

Type Syntax
~~~~~~~~~~~

At the level of terms, we simply elaborate on Ltac1 syntax, which is quite

> close to e.g. the one of OCaml. Types follow the simply-typed syntax of OCaml.

Reviewer: Well, it's better than OCaml, it has well-braced pattern-matching. Furthermore, the rationale was to keep as close as possible to Ltac1.

At the level of terms, we simply elaborate on Ltac1 syntax, which is quite
close to e.g. the one of OCaml. Types follow the simply-typed syntax of OCaml.

.. productionlist:: coq

>    type             : ( `type`, ... , `type` ) `typeconst`

Reviewer 3: Are the namespaces actually global?  Or intended to be global?  The `coq` after `productionlist` is meant to identify different sets of productions, which when I tried it seemed to correctly distinguish the different sets.

On the other hand, it will be very confusing for users to see multiple inconsistent productions for the same non-terminals in different chapters.  So I agree the names should be changed in doc.  Also, the grammar appears to repeat many productions that are already in Coq.

>    type             : ( `type`, ... , `type` ) `typeconst`

Reviewer 4: They are global in the sense that when you refer to an existing token through `:token:` or `:n:` or `:cmd:`, etc., then you have no way to say to which productionlist it belongs. I still have no idea what is the exact purpose of the qualifier after `productionlist`.

>    type             : ( `type`, ... , `type` ) `typeconst`

Reviewer 3: Defining the same nonterminal twice under "coq" gives an error; defining under separate labels does not.

But as you note, token doesn't work with this mechanism.

>    type             : ( `type`, ... , `type` ) `typeconst`

Author: > Defining the same nonterminal twice under "coq" gives an error; defining under separate labels does not.

@jfehrle then why didn't redefining `term` fail?

>    type             : ( `type`, ... , `type` ) `typeconst`

Reviewer 3: > then why didn't redefining `term` fail?

No idea.  I have seen duplicate nonterminal messages in the past but apparently they aren't generated in all the cases we would like.  Maybe the duplicate checking doesn't work across multiple doc files.  The Sphinx doc doesn't go into such detail.  Or it may be a Sphinx bug.

                    : ( `type` * ... * `type` )
                    : `type` -> `type`
                    : `typevar`
   typeconst        : ( `modpath` . )* `lident`

>    typevar          : '`lident`

Author: It's a lowercase ident, I believe. right, @ppedrot?

>    typevar          : '`lident`

Reviewer 4: OK then it can indeed be defined in the text with `:production:`.

>    typevar          : '`lident`

Reviewer: @maximedenes indeed, Ltac2 uses the lowercase / uppercase distinction similarly to OCaml.

   ltac2_typeparams       : ( `ltac2_typevar`, ... , `ltac2_typevar` )

The set of base types can be extended thanks to the usual ML type
declarations such as algebraic datatypes and records.

Built-in types include:

- ``int``, machine integers (size not specified, in practice inherited from OCaml)

> - ``string``, mutable strings

Reviewer: Ltac2 is still experimental, this kind of API can be discussed thoroughly via an early RFC (and ported quite easily, Ltac2 devs won't probably rely in an intricate way on this).

> - ``string``, mutable strings

Reviewer 4: OK, that's cool to hear.

- :n:`'a array`, mutable arrays
- :n:`exn`, exceptions
- :n:`constr`, kernel-side terms
- :n:`pattern`, term patterns

> - :n:`ident`, well-formed identifiers

Author: @Zimmi48 Ah right, what should I put?

> - :n:`ident`, well-formed identifiers

Reviewer 4: Double backquotes I would say.

Type declarations
~~~~~~~~~~~~~~~~~

One can define new types by the following commands.

.. cmd:: Ltac2 Type @typeparams @lident
   :name: Ltac2 Type

   This command defines an abstract type. It has no use for the end user and
   is dedicated to types representing data coming from the OCaml world.

.. cmdv:: Ltac2 Type {? rec} @typeparams @lident := @typedef

   This command defines a type with a manifest. There are four possible
   kinds of such definitions: alias, variant, record and open variant types.

   .. productionlist:: coq
      typedef          : `type`
                       : [ `constructordef` | ... | `constructordef` ]
                       : { `fielddef` ; ... ; `fielddef` }
                       : [ .. ]
      constructordef   : `ident` [ ( `type` , ... , `type` ) ]

>       fielddef         : [ mutable ] `ident` : `type`

Reviewer 4: Are these `ident` any different from `lident`?

   Aliases are just a name for a given type expression and are transparently
   unfoldable to it. They cannot be recursive.

   Variants are sum types defined by constructors and eliminated by

>    pattern-matching. They can be recursive, but the `RECFLAG` must be explicitly

Reviewer 4: `RECFLAG` or `rec` flag?

>    :production:`uident` designates identifiers starting with an uppercase.

Reviewer: Fixed.

>    :production:`uident` designates identifiers starting with an uppercase.

Reviewer 4: Maybe the fix is not pushed yet then.

>    :production:`uident` designates identifiers starting with an uppercase.

Reviewer: Yes, I've added a newline. Wasn't that what you meant?

>    :production:`uident` designates identifiers starting with an uppercase.

Author: @Zimmi48 are you ok with the current status of this?

>    :production:`uident` designates identifiers starting with an uppercase.

Reviewer 4: I don't see any newline. I was saying that I thought the sentence "The non-terminal :production:`uident` designates identifiers starting with an uppercase." would be better in the next paragraph. But this is not very important.

   pattern-matching. They can be recursive, but the `rec` flag must be
   explicitly set. Pattern-maching must be exhaustive.

   Records are product types with named fields and eliminated by projection.
   Likewise they can be recursive if the `rec` flag is set.

>    .. cmdv:: Ltac2 Type @ltac2_typeparams @qualid := [ @ltac2_constructordef ]

Reviewer: Indeed.

>    .. cmdv:: Ltac2 Type @ltac2_typeparams @qualid := [ @ltac2_constructordef ]

Reviewer: Fixed.

      Open variants are a special kind of variant types whose constructors are not
      statically defined, but can instead be extended dynamically. A typical example
      is the standard `exn` type. Pattern-matching must always include a catch-all
      clause. They can be extended by this command.

Term Syntax
~~~~~~~~~~~

The syntax of the functional fragment is very close to the one of Ltac1, except
that it adds a true pattern-matching feature, as well as a few standard
constructions from ML.

.. productionlist:: coq
   var              : lident
   qualid           : ( modpath . )* lident
   constructor      : uident

>    term             : qualid

Reviewer 4: There are some missing backquotes here and there is the same problem about namespacing.

>    ltac2_constructor: `uident`

Reviewer 4: If this synonymous has to be defined, why doing it here and not above and not in the piece of grammar above?

Reviewer: Which entry are you talking about? `ltac2_constructor`? In which case, it looks natural to put here, this is about Ltac2 terms.

Reviewer 4: Yes, and OK, if you feel like it belongs here. It's just that it is also part of the syntax of type definition.

   ltac2_term       : `ltac2_qualid`

>                     : `ltac2_constructor` `ltac2_term` ... `ltac2_term`

Reviewer 4: Can't constructor application just be normal function application, which allows things like `|> Some` (which are not allowed in OCaml, but are in other sane languages of the ML family)? In that case, it just means that a constructor is a term of the language.

Reviewer: That's already the case IIRC, and if not, it is compatible with the syntax exposed here anyways.

Reviewer 4: I would then just write:

```suggestion
                    : `ltac2_constructor`
```

Reviewer: Why not, will do.

                    : `ltac2_term` `ltac2_term` ... `ltac2_term`
                    : fun `ltac2_var` => `ltac2_term`
                    : let `ltac2_var` := `ltac2_term` in `ltac2_term`
                    : let rec `ltac2_var` := `ltac2_term` in `ltac2_term`
                    : match `ltac2_term` with `ltac2_branch` ... `ltac2_branch` end
                    : `int`

>                     : `string`

Reviewer 4: Why are `int` and `string` terms?

Author: Hum, they are just literals I guess. What's surprising?

Reviewer 4: Sorry, I don't get it. Since types are not terms in ML languages, I don't understand in which case this is useful. Furthermore, if `int` and `string` are terms, why not `constr` and `ident`?

Reviewer: I agree that they shouldn't be part of the syntax, they're already covered by literals. And this is not PHP (nor Ltac), they don't have any specific status parsingwise.

Author: @Zimmi48 I think you are misreading it. These are non-terminals.

@ppedrot I haven't checked if they were redundant, i.e. if literals were already included somewhere else. I haven't added these two, though, AFAIR, so it comes from the previous doc.

Reviewer: Ah! That's a misunderstanding actually. They stand for literal integers and strings, respectively. The original doc had probably some markup around.

Author: > Ah! That's a misunderstanding actually. They stand for literal integers and strings, respectively. The original doc had probably some markup around.

I'm lost. Isn't still the meaning they have in this doc?

Reviewer: I don't know what they're supposed to mean in the Sphinx-Coq grammar description language, but what I meant is that the intended meaning is indeed literals. What the correct way to write this down in the DSL is, I have no idea.

Reviewer 4: Literals and not non-terminals? Because indeed, Maxime's explanation did make sense to me (and corresponds indeed to the fact that there are backquotes around).

                    : [| `ltac2_term` ; ... ; `ltac2_term` |]
                    : ( `ltac2_term` , ... , `ltac2_term` )
                    : { `ltac2_field` `ltac2_field` ... `ltac2_field` }
                    : `ltac2_term` . ( `ltac2_qualid` )
                    : `ltac2_term` . ( `ltac2_qualid` ) := `ltac2_term`
                    : [; `ltac2_term` ; ... ; `ltac2_term` ]
                    : `ltac2_term` :: `ltac2_term`

>                     : ...

Reviewer 4: Is this supposed to mean that this grammar contains more things, or is this a syntactic construction?

Reviewer: More things.

   pattern          : `var`
                    : _
                    : ( `pattern` , ... , `pattern` )
                    : `constructor` `pattern` ... `pattern`
                    : [ ]
                    : `pattern` :: `pattern`
   field            : `qualid` := `term`

In practice, there is some additional syntactic sugar that allows e.g. to
bind a variable and match on it at the same time, in the usual ML style.

There is a dedicated syntax for list and array litterals.

> Limitations: for now, deep pattern matching is not implemented yet.

Reviewer 4: You could use the `.. note::` directive here.

> bind a variable and match on it at the same time, in the usual ML style.

Reviewer 4: An `as` clause? Why not include it here?

Reviewer: The syntax for pattern-matching is not fully specified yet, as it is missing deep patterns. I'll postpone its description until this is implemented.

.. note::

>    For now, deep pattern matching is not implemented yet.

Reviewer 4: Redundant. Say:
```suggestion
   For now, deep pattern matching is not implemented.
```

or:

```suggestion
   Deep pattern matching is not implemented yet.
```

Reviewer: Fixed.

   Mutable definitions act like dynamic binding, i.e. at runtime, the last defined
   value for this entry is chosen. This is useful for global flags and the like.

Reduction
~~~~~~~~~

We use the usual ML call-by-value reduction, with an otherwise unspecified
evaluation order. This is a design choice making it compatible with OCaml,
if ever we implement native compilation. The expected equations are as follows::

  (fun x => t) V ≡ t{x := V} (βv)

  let x := V in t ≡ t{x := V} (let)

  match C V₀ ... Vₙ with ... | C x₀ ... xₙ  => t | ... end ≡ t {xᵢ := Vᵢ} (ι)

  (t any term, V values, C constructor)

Note that call-by-value reduction is already a departure from Ltac1 which uses
heuristics to decide when evaluating an expression. For instance, the following
expressions do not evaluate the same way in Ltac1.

:n:`foo (idtac; let x := 0 in bar)`

:n:`foo (let x := 0 in bar)`

Instead of relying on the :n:`idtac` hack, we would now require an explicit thunk
not to compute the argument, and :n:`foo` would have e.g. type
:n:`(unit -> unit) -> unit`.

:n:`foo (fun () => let x := 0 in bar)`

Typing
~~~~~~

Typing is strict and follows Hindley-Milner system. We will not implement the

> current hackish subtyping semantics, and one will have to resort to conversion

Reviewer 4: Not clear what this means.

Reviewer: That means e.g. no magical casting from identifiers to constr and vice-versa. I am afraid that one has to understand Ltac1 semantics in depth to fully appreciate this sentence :/

Reviewer: Rephrased.

functions. See notations though to make things more palatable.

In this setting, all usual argument-free tactics have type :n:`unit -> unit`, but
one can return as well a value of type :n:`t` thanks to terms of type :n:`unit -> t`,
or take additional arguments.

Effects
~~~~~~~

> Regarding effects, nothing involved here, except that instead of using the

Reviewer 4: This language feels too colloquial.

Reviewer: Fixed.

  val throw : exn -> 'a

Contrarily to backtracking exceptions from the next section, this kind of error
is never caught by backtracking primitives, that is, throwing an exception
destroys the stack. This is materialized by the following equation, where `E`
is an evaluation context.::

  E[throw e] ≡ throw e

  (e value)

There is currently no way to catch such an exception and it is a design choice.
There might be at some future point a way to catch it in a brutal way,
destroying all backtrack and return values.

Backtrack
+++++++++

In Ltac2, we have the following backtracking primitives, defined in the
`Control` module.::

  Ltac2 Type 'a result := [ Val ('a) | Err (exn) ].

  val zero : exn -> 'a
  val plus : (unit -> 'a) -> (exn -> 'a) -> 'a
  val case : (unit -> 'a) -> ('a * (exn -> 'a)) result

If one sees thunks as lazy lists, then `zero` is the empty list and `plus` is
list concatenation, while `case` is pattern-matching.

The backtracking is first-class, i.e. one can write
:n:`plus (fun () => "x") (fun _ => "y") : string` producing a backtracking string.

These operations are expected to satisfy a few equations, most notably that they

> form a monoid compatible with sequentialization.::

Reviewer 4: Where is sequentialization defined? It is not in the term grammar. Do you intend people to refer to the definition from Ltac1?

Reviewer: This is an omission, the syntax should contain a node for `TERM; TERM`.

Reviewer: Sequentializations has been added to term syntax.

  plus t zero ≡ t ()
  plus (fun () => zero e) f ≡ f e
  plus (plus t f) g ≡ plus t (fun e => plus (f e) g)

  case (fun () => zero e) ≡ Err e

>   case (fun () => plus (fun () => t) f) ≡ Val t f

Reviewer 4: Isn't it

```suggestion
  case (fun () => plus (fun () => t) f) ≡ Val (t, f)
```

? `Val` is supposed to take a single argument.

Reviewer: This is an ADT constructor. Ltac2 uses the same syntax for applied constructors and functions, similarly to Gallina and unlike OCaml.

Reviewer 4: Maybe the issue is in the definition of the `result` type above but I see that the `Val` constructor only takes a single argument?

  (t, u, f, g, e values)

Goals
+++++

A goal is given by the data of its conclusion and hypotheses, i.e. it can be
represented as `[Γ ⊢ A]`.

The tactic monad naturally operates over the whole proofview, which may
represent several goals, including none. Thus, there is no such thing as
*the current goal*. Goals are naturally ordered, though.

It is natural to do the same in Ltac2, but we must provide a way to get access
to a given goal. This is the role of the `enter` primitive, that applies a
tactic to each currently focused goal in turn.::

  val enter : (unit -> unit) -> unit

It is guaranteed that when evaluating `enter f`, `f` is called with exactly one
goal under focus. Note that `f` may be called several times, or never, depending
on the number of goals under focus before the call to `enter`.

Accessing the goal data is then implicit in the Ltac2 primitives, and may panic
if the invariants are not respected. The two essential functions for observing
goals are given below.::

  val hyp : ident -> constr
  val goal : unit -> constr

The two above functions panic if there is not exactly one goal under focus.

> In addition, `hyp` may also fail if there is no hypothesis with the

Reviewer 4: panic or raise a catchable exception?

Reviewer: Failure means catchable exception. Anyways, this is not the right place to describe the functions part of the API, their precise behaviour is described in the `.v` files themselves, as is done in `mli` files for OCaml. The Ltac1 way of globally describing available tactics in the documentation is partly responsible for the namespace mess (or lack thereof). Because people expect to get tactics described in the documentation for free, this forces the plugins to define them as global notations.

or using the `ltac1val` quotation. The latter parses the same as `ltac1`, but
has type `Ltac2.Ltac1.t` instead of `unit`, and dynamically behaves as an Ltac1
thunk, i.e. `ltac1val:(foo)` corresponds to the tactic closure that Ltac1
would generate from `idtac; foo`.

Due to intricate dynamic semantics, understanding when Ltac1 value quotations
focus is very hard. This is why some functions return a continuation-passing
style value, as it can dispatch dynamically between focused and unfocused
behaviour.

Ltac2 from Ltac1
~~~~~~~~~~~~~~~~

Same as above by switching Ltac1 by Ltac2 and using the `ltac2` quotation
instead.

Note that the tactic expression is evaluated eagerly, if one wants to use it as
an argument to a Ltac1 function, she has to resort to the good old
:n:`idtac; ltac2:(foo)` trick. For instance, the code below will fail immediately
and won't print anything.

.. coqtop:: in

   From Ltac2 Require Import Ltac2.
   Set Default Proof Mode "Classic".

.. coqtop:: all

   Ltac mytac tac := idtac "wow"; tac.

   Goal True.
   Proof.
   Fail mytac ltac2:(fail).

Transition from Ltac1
---------------------

Owing to the use of a bunch of notations, the transition shouldn't be

> atrociously horrible and shockingly painful up to the point you want to retire

Reviewer 3: "atrociously horrible and shockingly painful etc. etc." - really??  While hyperbolic, it's also very vague.

Tactic delay
~~~~~~~~~~~~

Tactics are not magically delayed anymore, neither as functions nor as
arguments. It is your responsibility to thunk them beforehand and apply them
at the call site.

A typical example of a delayed function:

:n:`Ltac foo := blah.`

becomes

:n:`Ltac2 foo () := blah.`

All subsequent calls to `foo` must be applied to perform the same effect as
before.

Likewise, for arguments:

:n:`Ltac bar tac := tac; tac; tac.`

becomes

:n:`Ltac2 bar tac := tac (); tac (); tac ().`

We recommend the use of syntactic notations to ease the transition. For
instance, the first example can alternatively written as:

:n:`Ltac2 foo0 () := blah.`
:n:`Ltac2 Notation foo := foo0 ().`

This allows to keep the subsequent calls to the tactic as-is, as the
expression `foo` will be implicitly expanded everywhere into `foo0 ()`. Such
a trick also works for arguments, as arguments of syntactic notations are
implicitly thunked. The second example could thus be written as follows.

:n:`Ltac2 bar0 tac := tac (); tac (); tac ().`
:n:`Ltac2 Notation bar := bar0.`

Variable binding
~~~~~~~~~~~~~~~~

Ltac1 relies on a crazy amount of dynamic trickery to be able to tell apart

> bound variables from terms, hypotheses and whatnot. There is no such thing in

Reviewer 3: crazy??

Reviewer: You've never read the Ltac implementation, have you? :) But, yep let's make this neutral.

.. _ltac2:

Ltac2
=====

The Ltac tactic language is probably one of the ingredients of the success of
Coq, yet it is at the same time its Achilles' heel. Indeed, Ltac:

- has often unclear semantics
- is very non-uniform due to organic growth
- lacks expressivity (data structures, combinators, types, ...)
- is slow
- is error-prone and fragile
- has an intricate implementation

Following the need of users that start developing huge projects relying
critically on Ltac, we believe that we should offer a proper modern language
that features at least the following:

- at least informal, predictable semantics
- a typing system
- standard programming facilities (i.e. datatypes)

This new language, called Ltac2, is described in this chapter. It is still
experimental but we encourage nonetheless users to start testing it,
especially wherever an advanced tactic language is needed. The previous

> implementation of Ltac, described in the previous chapter, will be referred to

Reviewer: Experimental means that the APIs are prone to change without previous notice. I don't know whether α encompasses this kind of feature.

> We should have a much more positive introduction to the chapter to motivate users to begin using the feature

I am quite confident that Ltac1 is a motivation strong enough to switch to Ltac2, but I am obviously biased...

> implementation of Ltac, described in the previous chapter, will be referred to

Reviewer 3: Then we should say in the text "the APIs are subject to change without notice" but also state that the goal is to eventually retire Ltac1.  And presumably API changes would be primarily for bugs, conversion problems and design issues.

Maybe also say that we'd like to work with users who are willing to try to convert existing Ltac1 scripts--meaning someone would make an effort to be responsive to questions/issues and make fixes on a timely basis (perhaps quicker than the next official release; maybe a branch).  If that's something you or others are willing to undertake.  Otherwise I can imagine a bunch of users running into the same issues repeatedly, a waste of time for both users and developers.  Also less likely for the new code to possibly get a bad reputation due to early issues.  Also, without this, the transition time to get the community onto Ltac2 could be many years.

> I am quite confident that Ltac1 is a motivation strong enough to switch to Ltac2, but I am obviously biased...

How about new users?  The Ltac1 documentation, poor as it is, is more complete than the Ltac2 doc.  I'm not sure what choice I would make.  (How about people who gave up on Ltac1?)

Also, are there important proof libraries that use Ltac1 in a way that prevents using them with Ltac2?  If so, the incompatibility would block their users from using Ltac2.  It might be worth some time to identify these libraries and their interdependencies.  (That might identify good candidates for the trial conversions I suggested above.)  Also, I recall seeing info on calling Ltac1 from Ltac2 but I don't remember seeing info on calling Ltac2 from Ltac1.  Are the restrictions documented?

Of course, this all depends on what people are willing to spend their time on.

> implementation of Ltac, described in the previous chapter, will be referred to

Reviewer: I am postponing the change of this paragraph after the merge. This kind of introductory text has to be written before the next release, and needs a bit of serious pondering.

> implementation of Ltac, described in the previous chapter, will be referred to

Reviewer 3: Sounds good.  If you wish, I'd be happy to review the change and/or help with wording.

as Ltac1.

.. _ltac2_design:

General design
--------------

There are various alternatives to Ltac1, such that Mtac or Rtac for instance.
While those alternatives can be quite distinct from Ltac1, we designed
Ltac2 to be closest as reasonably possible to Ltac1, while fixing the
aforementioned defects.

In particular, Ltac2 is:

- a member of the ML family of languages, i.e.

  * a call-by-value functional language
  * with effects
  * together with Hindley-Milner type system

- a language featuring meta-programming facilities for the manipulation of
  Coq-side terms
- a language featuring notation facilities to help writing palatable scripts

We describe more in details each point in the remainder of this document.

ML component
------------

Overview
~~~~~~~~

Ltac2 is a member of the ML family of languages, in the sense that it is an
effectful call-by-value functional language, with static typing à la
Hindley-Milner (see :cite:`MilnerPrincipalTypeSchemes`). It is commonly accepted
that ML constitutes a sweet spot in PL design, as it is relatively expressive

> while not being either too lax (contrarily to dynamic typing) nor too strict

Reviewer 3: ```suggestion
while not being either too lax (unlike dynamic typing) nor too strict
```

Not grammatical.

> (contrarily to say, dependent types).

Reviewer 3: ```suggestion
(unlike, say, dependent types).
```

   If ``mutable`` is set, the definition can be redefined at a later stage (see below).

.. cmd:: Ltac2 Set @qualid := @ltac2_term
   :name: Ltac2 Set

   This command redefines a previous ``mutable`` definition.
   Mutable definitions act like dynamic binding, i.e. at runtime, the last defined
   value for this entry is chosen. This is useful for global flags and the like.

Reduction
~~~~~~~~~

We use the usual ML call-by-value reduction, with an otherwise unspecified
evaluation order. This is a design choice making it compatible with OCaml,
if ever we implement native compilation. The expected equations are as follows::

  (fun x => t) V ≡ t{x := V} (βv)

  let x := V in t ≡ t{x := V} (let)

  match C V₀ ... Vₙ with ... | C x₀ ... xₙ  => t | ... end ≡ t {xᵢ := Vᵢ} (ι)

  (t any term, V values, C constructor)

Note that call-by-value reduction is already a departure from Ltac1 which uses
heuristics to decide when evaluating an expression. For instance, the following
expressions do not evaluate the same way in Ltac1.

:n:`foo (idtac; let x := 0 in bar)`

:n:`foo (let x := 0 in bar)`

Instead of relying on the :n:`idtac` hack, we would now require an explicit thunk

> not to compute the argument, and :n:`foo` would have e.g. type

Reviewer 3: hack??

Reviewer: Writing `idtac; t` is the "idiomatic" way to delay the evaluation of  `t` in Ltac, but it doesn't work for value-returning tactics. So it's a hack. The whole semantics of Ltac is based on an undecipherable dynamic heuristic to dispatch between by-value and by-name.

Reviewer 3: Sure.  My point was that "hack" is often a pejorative term.

:n:`(unit -> unit) -> unit`.

:n:`foo (fun () => let x := 0 in bar)`

Typing
~~~~~~

> Typing is strict and follows Hindley-Milner system. Contrarily to Ltac1, there

Reviewer 3: ```suggestion
Typing is strict and follows Hindley-Milner system. Unlike Ltac1, there
```

- It can fail in a non-recoverable way
- It can use first-class backtrack. The proper way to figure that is that we
  morally have the following isomorphism:
  :n:`(unit -> 'a) ~ (unit -> exn + ('a * (exn -> 'a)))`
  i.e. thunks can produce a lazy list of results where each
  tail is waiting for a continuation exception.
- It can access a backtracking proof state, made out amongst other things of
  the current evar assignation and the list of goals under focus.

We describe more thoroughly the various effects existing in Ltac2 hereafter.

Standard IO
+++++++++++

The Ltac2 language features non-backtracking IO, notably mutable data and
printing operations.

Mutable fields of records can be modified using the set syntax. Likewise,
built-in types like `string` and `array` feature imperative assignment. See
modules `String` and `Array` respectively.

A few printing primitives are provided in the `Message` module, allowing to
display information to the user.

Fatal errors
++++++++++++

The Ltac2 language provides non-backtracking exceptions through the
following primitive in module `Control`.::

  val throw : exn -> 'a

> Contrarily to backtracking exceptions from the next section, this kind of error

Reviewer 3: ```suggestion
Unlike backtracking exceptions from the next section, this kind of error
```

  let x := zero e in u ≡ zero e
  let x := plus t f in u ≡ plus (fun () => let x := t in u) (fun e => let x := f e in u)

  (t, u, f, g, e values)

Goals
+++++

A goal is given by the data of its conclusion and hypotheses, i.e. it can be
represented as `[Γ ⊢ A]`.

The tactic monad naturally operates over the whole proofview, which may
represent several goals, including none. Thus, there is no such thing as
*the current goal*. Goals are naturally ordered, though.

It is natural to do the same in Ltac2, but we must provide a way to get access
to a given goal. This is the role of the `enter` primitive, that applies a
tactic to each currently focused goal in turn.::

  val enter : (unit -> unit) -> unit

It is guaranteed that when evaluating `enter f`, `f` is called with exactly one
goal under focus. Note that `f` may be called several times, or never, depending
on the number of goals under focus before the call to `enter`.

Accessing the goal data is then implicit in the Ltac2 primitives, and may panic
if the invariants are not respected. The two essential functions for observing
goals are given below.::

  val hyp : ident -> constr
  val goal : unit -> constr

> The two above functions panic if there is not exactly one goal under focus.

Reviewer 3: "panic"?  I think you mean "fail" or "generate an exception".  In the Linux world, "panic" means the OS is going to halt with an error condition.

Reviewer: That's a terminology I borrowed from Rust. I think the documentation describes it before, but FTR there are two kinds of exceptions in Ltac2, backtracking exceptions (that can be used for control) and panics, that are not catchable and lead to a toplevel error.

Reviewer 3: "Uncatchable exception" would be better then.  "panic" is not a common usage (nor is Rust widely used?), if you stick with "panic" the documentation should define the term.

Reviewer: I described the "panic" terminology in the doc.

In addition, `hyp` may also fail if there is no hypothesis with the
corresponding name.

Meta-programming
----------------

Overview
~~~~~~~~

> One of the horrendous implementation issues of Ltac is the fact that it is

Reviewer 3: ```suggestion
One of the major implementation issues of Ltac is the fact that it is
```

never clear whether an object refers to the object world or the meta-world.
This is an incredible source of slowness, as the interpretation must be
aware of bound variables and must use heuristics to decide whether a variable
is a proper one or referring to something in the Ltac context.

Likewise, in Ltac1, constr parsing is implicit, so that ``foo 0`` is
not ``foo`` applied to the Ltac integer expression ``0`` (Ltac does have a
non-first-class notion of integers), but rather the Coq term :g:`Datatypes.O`.

> We should stop doing that! We need to mark when quoting and unquoting, although

Reviewer 3: ```suggestion
The implicit parsing is confusing to users and often gives unexpected results.  Ltac2 makes these explicit using quoting and unquoting notation, although
```

insert in a concise way an Ltac2 variable of type :n:`constr` into a Coq term.

Match over terms
~~~~~~~~~~~~~~~~

Ltac2 features a construction similar to Ltac1 :n:`match` over terms, although
in a less hard-wired way.

.. productionlist:: coq
   ltac2_term       : match! `ltac2_term` with `constrmatching` .. `constrmatching` end
                    : lazy_match! `ltac2_term` with `constrmatching` .. `constrmatching` end
                    : multi_match! `ltac2_term` with `constrmatching` .. `constrmatching` end
   constrmatching  : | `constrpattern` => `ltac2_term`
   constrpattern   : `term`
                   : context  [ `term` ]
                   : context `ident` [ `term` ]

This construction is not primitive and is desugared at parsing time into
calls to term matching functions from the `Pattern` module. Internally, it is
implemented thanks to a specific scope accepting the :n:`@constrmatching` syntax.

Variables from the :n:`@constrpattern` are statically bound in the body of the branch, to
values of type `constr` for the variables from the :n:`@constr` pattern and to a
value of type `Pattern.context` for the variable :n:`@lident`.

> Note that contrarily to Ltac, only lowercase identifiers are valid as Ltac2

Reviewer 3: ```suggestion
Note that unlike Ltac, only lowercase identifiers are valid as Ltac2
```

The original is not grammatical; "contrarily" is an adverb.

errors after expansion.

Evaluation
----------

Ltac2 features a toplevel loop that can be used to evaluate expressions.

.. cmd:: Ltac2 Eval @term
   :name: Ltac2 Eval

   This command evaluates the term in the current proof if there is one, or in the
   global environment otherwise, and displays the resulting value to the user
   together with its type. This function is pure in the sense that it does not
   modify the state of the proof, and in particular all side-effects are discarded.

Debug
-----

.. opt:: Ltac2 Backtrace

   When this option is set, toplevel failures will be printed with a backtrace.

Compatibility layer with Ltac1
------------------------------

Ltac1 from Ltac2
~~~~~~~~~~~~~~~~

Simple API
++++++++++

One can call Ltac1 code from Ltac2 by using the :n:`ltac1` quotation. It parses
a Ltac1 expression, and semantics of this quotation is the evaluation of the
corresponding code for its side effects. In particular, in cannot return values,
and the quotation has type :n:`unit`.

Beware, Ltac1 **cannot** access variables from the Ltac2 scope. One is limited
to the use of standalone function calls.

Low-level API
+++++++++++++

> There exists a lower-level FFI into Ltac1 that is not recommended for daily use,

Reviewer 3: ```suggestion
There exists a lower-level API into Ltac1 that is not recommended for daily use,
```

Reviewer: Why not FFI? That's the designated nomenclature for a trans-language API.

Reviewer 3: I thought it was a typo.  Even so, it's probably not a term that most readers would be familiar with.  This is the only use in the document, it's use is not essential here (compared to "API") and there's no definition.

Documentation should be written to be simple to understand, not introducing additional terms unless it's really necessary.  Terms that may not be understood by, say, half the readers should be avoided or defined.  I think that's different from academic writing where use of the latest jargon is probably expected.

Author: FFI is by no means academic IMHO, even JavaScript programmers know the term :) It sounds as standard as API to me.

Reviewer 3: I didn't say it's an academic term.  I never ran across it when I've written JavaScript.  How many of our users know JavaScript and would know the term?

Reviewer: It's not specific to JavaScript, @maximedenes mentioned this as an example. *FFI* is a term whose usage frequency is on par with *API*, for the good reason that they are notions very similar in nature.

Reviewer 3: Our opinions are subjective.  I've worked in Silicon Valley a long while and never encountered the term.  Terminology choices vary across groups and other categories.  Let's not get hung up on this now.

Of course, "JavaScript programmers" is not equivalent to "inexperienced/junior programmers".  In some ways JavaScript is harder to use than OCaml (e.g. JS has minimal type checking).  Delivering significant applications that work across multiple browsers can be quite complex.

Reviewer 5: "FFI" is common, I believe, when developers use two languages that have to interface, and more common if those two languages are at different places in the high-level / low-level spectrum.  So you might talk about a C FFI for OCaml/Haskell/Python/Perl/etc, or a Python FFI for Perl, or a Lua FFI for TeX.

According to [Wikipedia](https://en.wikipedia.org/wiki/Foreign_function_interface):
> The term comes from the specification for Common Lisp, which explicitly refers to the language features for inter-language calls as such;[1] the term is also used officially by the Haskell[2] and Python programming languages.[3] Other languages use other terminology: the Ada programming language talks about "language bindings", while Java refers to its FFI as the JNI (Java Native Interface) or JNA (Java Native Access). Foreign function interface has become generic terminology for mechanisms which provide such services.

Author: > Of course, "JavaScript programmers" is not equivalent to "inexperienced/junior programmers".

FTR, I never meant that.

The Ltac2 language features non-backtracking IO, notably mutable data and
printing operations.

Mutable fields of records can be modified using the set syntax. Likewise,
built-in types like `string` and `array` feature imperative assignment. See
modules `String` and `Array` respectively.

A few printing primitives are provided in the `Message` module, allowing to
display information to the user.

Fatal errors
++++++++++++

The Ltac2 language provides non-backtracking exceptions, also known as *panics*,
through the following primitive in module `Control`.::

  val throw : exn -> 'a

Unlike backtracking exceptions from the next section, this kind of error
is never caught by backtracking primitives, that is, throwing an exception
destroys the stack. This is materialized by the following equation, where `E`
is an evaluation context.::

  E[throw e] ≡ throw e

  (e value)

There is currently no way to catch such an exception and it is a design choice.
There might be at some future point a way to catch it in a brutal way,
destroying all backtrack and return values.

Backtrack
+++++++++

In Ltac2, we have the following backtracking primitives, defined in the
`Control` module.::

>   Ltac2 Type 'a result := [ Val ('a) | Err (exn) ].

Reviewer 4: Let me insist on this point: here it seems that constructor `Val` only takes a single argument, but later on it is used with two arguments:
```
case (fun () => plus (fun () => t) f) ≡ Val t f
```

Reviewer: It seems you're actually right, in the Ltac2 source `Val` is unary. The type tells us that the observation function produces a `Val` containing a pair of type `'a * (exn -> 'a)`, so this should be `Val (t, f)` instead.

It is guaranteed that when evaluating `enter f`, `f` is called with exactly one
goal under focus. Note that `f` may be called several times, or never, depending
on the number of goals under focus before the call to `enter`.

Accessing the goal data is then implicit in the Ltac2 primitives, and may panic
if the invariants are not respected. The two essential functions for observing
goals are given below.::

  val hyp : ident -> constr
  val goal : unit -> constr

The two above functions panic if there is not exactly one goal under focus.
In addition, `hyp` may also fail if there is no hypothesis with the
corresponding name.

Meta-programming
----------------

Overview
~~~~~~~~

One of the major implementation issues of Ltac is the fact that it is
never clear whether an object refers to the object world or the meta-world.
This is an incredible source of slowness, as the interpretation must be
aware of bound variables and must use heuristics to decide whether a variable
is a proper one or referring to something in the Ltac context.

Likewise, in Ltac1, constr parsing is implicit, so that ``foo 0`` is
not ``foo`` applied to the Ltac integer expression ``0`` (Ltac does have a

> non-first-class notion of integers), but rather the Coq term :g:`Datatypes.O`.

Reviewer: This doesn't mean the same, though.

> non-first-class notion of integers), but rather the Coq term :g:`Datatypes.O`.

Reviewer 4: Yes, I am aware of this. But I'm not sure the distinction is worth the complication. But do as you prefer.

> non-first-class notion of integers), but rather the Coq term :g:`Datatypes.O`.

Reviewer 5: I believe a better phrasing would be "Ltac does have a notion of integer, but it is not first-class"

> non-first-class notion of integers), but rather the Coq term :g:`Datatypes.O`.

Reviewer 5: I believe a better phrasing would be "Ltac does have a notion of integer, but it is not first-class"

The implicit parsing is confusing to users and often gives unexpected results.
Ltac2 makes these explicit using quoting and unquoting notation, although

> we need to do that in a short and elegant way so as not to be too cumbersome

Reviewer 4: "we need to do that" -> "there are notations to do this"

to the user.

Generic Syntax for Quotations
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> In general, quotations can be introduced in term by the following syntax, where

Reviewer 4: ```suggestion
In general, quotations can be introduced in terms using the following syntax, where
```

> :n:`@quotentry` is some parsing entry.

Reviewer 4: ```suggestion
:production:`quotentry` is some parsing entry.
```

.. prodn::

>    term += @ident : ( @quotentry )

Reviewer 4: ```suggestion
   ltac2_term += @ident : ( @quotentry )
```

And `ident` itself is strange given that it is just meant to represent a short list of possible keywords...

Author: @ppedrot any comment on this?

Reviewer: > a short list of possible keywords

Not necessarily, the midterm goal is to make this mechanism extensible.

Built-in quotations
+++++++++++++++++++

The current implementation recognizes the following built-in quotations:

- ``ident``, which parses identifiers (type ``Init.ident``).
- ``constr``, which parses Coq terms and produces an-evar free term at runtime
  (type ``Init.constr``).
- ``open_constr``, which parses Coq terms and produces a term potentially with
  holes at runtime (type ``Init.constr`` as well).
- ``pattern``, which parses Coq patterns and produces a pattern used for term
  matching (type ``Init.pattern``).

> - ``reference``, which parses either a :n:`qualid` or :n:`& @ident`. Qualified names

Reviewer 4: ```suggestion
- ``reference``, which parses either a :n:`@qualid` or :n:`& @ident`. Qualified names
```

modes, respectively called the *strict* and the *non-strict* mode.

- In strict mode, all simple identifiers appearing in a term quotation are
  required to be resolvable statically. That is, they must be the short name of
  a declaration which is defined globally, excluding section variables and
  hypotheses. If this doesn't hold, internalization will fail. To work around
  this error, one has to specifically use the ``&`` notation.
- In non-strict mode, any simple identifier appearing in a term quotation which
  is not bound in the global context is turned into a dynamic reference to a
  hypothesis. That is to say, internalization will succeed, but the evaluation
  of the term at runtime will fail if there is no such variable in the dynamic
  context.

Strict mode is enforced by default, e.g. for all Ltac2 definitions. Non-strict
mode is only set when evaluating Ltac2 snippets in interactive proof mode. The
rationale is that it is cumbersome to explicitly add ``&`` interactively, while it
is expected that global tactics enforce more invariants on their code.

Term Antiquotations
~~~~~~~~~~~~~~~~~~~

Syntax
++++++

> One can also insert Ltac2 code into Coq term, similarly to what was possible in

Reviewer 4: ```suggestion
One can also insert Ltac2 code into Coq terms, similarly to what is possible in
```

Ltac1.

.. prodn::
   term += ltac2:( @ltac2_term )

Antiquoted terms are expected to have type ``unit``, as they are only evaluated
for their side-effects.

Semantics
+++++++++

Interpretation of a quoted Coq term is done in two phases, internalization and
evaluation.

- Internalization is part of the static semantics, i.e. it is done at Ltac2
  typing time.
- Evaluation is part of the dynamic semantics, i.e. it is done when
  a term gets effectively computed by Ltac2.

> Remark that typing of Coq terms is a *dynamic* process occuring at Ltac2

Reviewer 4: ```suggestion
Note that typing of Coq terms is a *dynamic* process occurring at Ltac2
```

evaluation time, and not at Ltac2 typing time.

Static semantics
****************

During internalization, Coq variables are resolved and antiquotations are
type-checked as Ltac2 terms, effectively producing a ``glob_constr`` in Coq
implementation terminology. Note that although it went through the
type-checking of **Ltac2**, the resulting term has not been fully computed and

> is potentially ill-typed as a runtime **Coq** term.::

Reviewer 4: You need to insert a newline everywhere where you put `.::` because `::` at the end of a line is transformed into `:` in the text:

```suggestion
is potentially ill-typed as a runtime **Coq** term.

::
```

or

```suggestion
is potentially ill-typed as a runtime **Coq** term::
```

  Ltac2 Definition myconstr () := constr:(nat -> 0).

>   // Valid with type `unit -> constr`, but will fail at runtime.

Reviewer 4: You could actually use a `coqdoc` or `coqtop` block here. It also seems very strange to use `//` to mark comments.

Author: @ppedrot does `Ltac Definition` really exists? I get a syntax error.

Reviewer 4: Indeed, this is probably a mistake and it was meant `Ltac2 myconstr ()` instead.

Reviewer: Indeed.

Term antiquotations are type-checked in the enclosing Ltac2 typing context
of the corresponding term expression. For instance, the following will
type-check.::

  let x := '0 in constr:(1 + ltac2:(exact x))
  // type constr

> Beware that the typing environment of typing of antiquotations is **not**

Reviewer 4: ```suggestion
Beware that the typing environment of antiquotations is **not**
```

expanded by the Coq binders from the term. Namely, it means that the following
Ltac2 expression will **not** type-check.::

  constr:(fun x : nat => ltac2:(exact x))
  // Error: Unbound variable 'x'

There is a simple reason for that, which is that the following expression would
not make sense in general.

> :n:`constr:(fun x : nat => ltac2:(clear @x; exact x))`

Reviewer 4: This `@` is part of the Ltac syntax. Why are you putting this into a `:n:` quotation that interprets `@` as starting a token?

Reviewer: @maximedenes? He's the one who ported the doc.

Indeed, a hypothesis can suddenly disappear from the runtime context if some
other tactic pulls the rug from under you.

Rather, the tactic writer has to resort to the **dynamic** goal environment,
and must write instead explicitly that she is accessing a hypothesis, typically
as follows.

> :n:`constr:(fun x : nat => ltac2:(exact (hyp @x)))`

Reviewer 4: Same.

This pattern is so common that we provide dedicated Ltac2 and Coq term notations
for it.

- `&x` as an Ltac2 expression expands to `hyp @x`.
- `&x` as a Coq constr expression expands to
  `ltac2:(Control.refine (fun () => hyp @x))`.

Dynamic semantics
*****************

During evaluation, a quoted term is fully evaluated to a kernel term, and is
in particular type-checked in the current environment.

Evaluation of a quoted term goes as follows.

- The quoted term is first evaluated by the pretyper.
- Antiquotations are then evaluated in a context where there is exactly one goal
  under focus, with the hypotheses coming from the current environment extended
  with the bound variables of the term, and the resulting term is fed into the
  quoted term.

Relative orders of evaluation of antiquotations and quoted term are not
specified.

For instance, in the following example, `tac` will be evaluated in a context
with exactly one goal under focus, whose last hypothesis is `H : nat`. The

> whole expression will thus evaluate to the term `fun H : nat => nat`.

Reviewer 4: Didn't you mean

```suggestion
whole expression will thus evaluate to the term :g:`fun H : nat => H`.
```

?

> :n:`let tac () := hyp @H in constr:(fun H : nat => ltac2:(tac ()))`

Reviewer 4: Same.

Many standard tactics perform type-checking of their argument before going
further. It is your duty to ensure that terms are well-typed when calling
such tactics. Failure to do so will result in non-recoverable exceptions.

> Trivial Term Antiquotations

Reviewer 4: ```suggestion
**Trivial Term Antiquotations**
```

It is possible to refer to a variable of type `constr` in the Ltac2 environment
through a specific syntax consistent with the antiquotations presented in
the notation section.

.. prodn:: term += $@lident

In a Coq term, writing :g:`$x` is semantically equivalent to
:g:`ltac2:(Control.refine (fun () => x))`, up to re-typechecking. It allows to
insert in a concise way an Ltac2 variable of type :n:`constr` into a Coq term.

Match over terms
~~~~~~~~~~~~~~~~

Ltac2 features a construction similar to Ltac1 :n:`match` over terms, although
in a less hard-wired way.

.. productionlist:: coq

>    ltac2_term       : match! `ltac2_term` with `constrmatching` .. `constrmatching` end

Reviewer 4: I think `productionlist` is not supposed to be used to extend preexisting grammar tokens and that `prodn` should be used instead.

Author: But this is a whole set of rules. `prodn` represents only one non-terminal, AFAIK.

                    : lazy_match! `ltac2_term` with `constrmatching` .. `constrmatching` end
                    : multi_match! `ltac2_term` with `constrmatching` .. `constrmatching` end
   constrmatching  : | `constrpattern` => `ltac2_term`
   constrpattern   : `term`
                   : context  [ `term` ]
                   : context `ident` [ `term` ]

This construction is not primitive and is desugared at parsing time into
calls to term matching functions from the `Pattern` module. Internally, it is
implemented thanks to a specific scope accepting the :n:`@constrmatching` syntax.

Variables from the :n:`@constrpattern` are statically bound in the body of the branch, to
values of type `constr` for the variables from the :n:`@constr` pattern and to a

> value of type `Pattern.context` for the variable :n:`@lident`.

Reviewer 4: Should it be `ident` here or `lident` above?

Author: @ppedrot do you know?

Reviewer: `lident` everywhere, these are Ltac2 variables.

Note that unlike Ltac, only lowercase identifiers are valid as Ltac2
bindings, so that there will be a syntax error if one of the bound variables
starts with an uppercase character.

The semantics of this construction is otherwise the same as the corresponding

> one from Ltac1, except that it requires the goal to be focused.

Reviewer 4: This would really benefit from an example. BTW, examples should generally be wrapped in a `.. example::` directive: this would make them easier to notice.

Author: I added `example` in some places. As for more examples, I hope this can be addressed later, if needed.

Match over goals
~~~~~~~~~~~~~~~~

Similarly, there is a way to match over goals in an elegant way, which is
just a notation desugared at parsing time.

.. productionlist:: coq

>    ltac2_term             : match! [ reverse ] goal with `goalmatching` ... `goalmatching` end

Reviewer 4: Same

                    : lazy_match! [ reverse ] goal with `goalmatching` ... `goalmatching` end
                    : multi_match! [ reverse ] goal with `goalmatching` ... `goalmatching` end
   goalmatching     : | [ `hypmatching` ... `hypmatching` |- `constrpattern` ] => `ltac2_term`

>    hypmatching      : `lident` : `constrpattern`

Reviewer 4: Can't it be `_` as well instead of just `lident`?

Author: @ppedrot do you know?

Reviewer: Probably, looking at the code.

Reviewer: I have no idea on the syntax to write the down though. I can probably fiddle with the doc generation until this works, but it's probably faster for you to put it here.

Reviewer 4: ```suggestion
   hypmatching      : `lident` : `constrpattern`
                    : _ : `constrpattern`
```

Variables from :n:`@hypmatching` and :n:`@constrpattern` are bound in the body of the
branch. Their types are:

- ``constr`` for pattern variables appearing in a :n:`@constr`
- ``Pattern.context`` for variables binding a context
- ``ident`` for variables binding a hypothesis name.

The same identifier caveat as in the case of matching over constr applies, and

> this features has the same semantics as in Ltac1. In particular, a ``reverse``

Reviewer 4: You can't count on this alone for people to understand Ltac2's lazymatch and multimatch. It should be possible for someone to learn about Ltac2 without an advanced knowledge of Ltac1.

Reviewer: Specifying the semantics of those commands require in-depth understanding of the backtracking model... The Ltac1 description is notoriously nonsense, in particular regarding `lazymatch`. These constructions are not primitive in Ltac2, only the "match one goal pattern with a continuation" is, the rest is given by Ltac2 code. Can't we just write that down?

Author: Not sure how to address this indeed. Documentation in the `.v` files I imagine.

flag can be specified to match hypotheses from the more recently introduced to
the least recently introduced one.

Notations
---------

Notations are the crux of the usability of Ltac1. We should be able to recover

> a feeling similar to the old implementation by using and abusing notations.

Reviewer 4: What do you mean?

Reviewer: That nothing can't be solved with macros?

Reviewer 4: To start with, did you really mean "Ltac1" and not "Ltac2" here?

Reviewer: Nope, that's a statement about Ltac1, as shown by the adjective "old".

Scopes
~~~~~~

A scope is a name given to a grammar entry used to produce some Ltac2 expression
at parsing time. Scopes are described using a form of S-expression.

.. prodn::
   scope ::= @string %| @int %| @lident ({+, @scope})

A few scopes contain antiquotation features. For sake of uniformity, all

> antiquotations are introduced by the syntax :n:`$@var`.

Author: @ppedrot ?

The following scopes are built-in.

- :n:`constr`:

  + parses :n:`c = @constr` and produces :n:`constr:(c)`

- :n:`ident`:

  + parses :n:`id = @ident` and produces :n:`ident:(id)`
  + parses :n:`$(x = @ident)` and produces the variable :n:`x`

- :n:`list0(@scope)`:

  + if :n:`@scope` parses :n:`@entry`, parses :n:`(x₀, ..., xₙ = {* @entry})` and produces
    :n:`[x₀; ...; xₙ]`.

- :n:`list0(@scope, sep = @string)`:

>   + if :n:`@scope` parses :n:`@entry`, parses `(x₀ = @entry, sep, ..., sep, xₙ = @entry)`

Author: @ppedrot ?

>   + if :n:`@scope` parses :n:`@entry`, parses `(x₀ = @entry, sep, ..., sep, xₙ = @entry)`

Reviewer: That's a meta-symbol, if that's what you mean. The entry parses a list of `entry` separated by `sep`, I have no idea of the Sphinx syntax for that.

>   + if :n:`@scope` parses :n:`@entry`, parses `(x₀ = @entry, sep, ..., sep, xₙ = @entry)`

Reviewer 4: If `sep` was a single character `s`, you could write:

```rst
:n:`{+s @entry }
```

with the caveat that `entry` is still not a grammar token.

>   + if :n:`@scope` parses :n:`@entry`, parses `(x₀ = @entry, sep, ..., sep, xₙ = @entry)`

Author: I made entry be defined locally, that's what the sentence means. I'll remove the comas.

>   + if :n:`@scope` parses :n:`@entry`, parses `(x₀ = @entry, sep, ..., sep, xₙ = @entry)`

Reviewer 3: > If `sep` was a single character `s`, you could write:

```rst
:n:`{+s @entry }`
```

IIRC multiple characters also work here.  The graphic in the output doesn't adjust fully to the longer string, but I think "sep" is short enough to be OK.

    and produces `[x₀; ...; xₙ]`.

- :n:`list1`: same as :n:`list0` (with or without separator) but parses :n:`{+ @entry}` instead
  of `ENTRY*`.

- :n:`opt(@scope)`

  + if :n:`@scope` parses :n:`@entry`, parses :n:`{? @entry}` and produces either :n:`None` or
    :n:`Some x` where :n:`x` is the parsed expression.

- :n:`self`:

  + parses a Ltac2 expression at the current level and return it as is.

- :n:`next`:

  + parses a Ltac2 expression at the next level and return it as is.

- :n:`tactic(n = @int)`:

  + parses a Ltac2 expression at the provided level :n:`n` and return it as is.

- :n:`thunk(@scope)`:

  + parses the same as :n:`scope`, and if :n:`e` is the parsed expression, returns
    :n:`fun () => e`.

- :n:`STRING`:

>   + parses the corresponding string as a CAMLP5 IDENT and returns :n:`()`.

Reviewer 4: What does this mean?

Reviewer: Not sure I understand the question, but e.g. writing "foo" will parse the identifier `foo`, and the result will be bound to the unique value of type unit.

Reviewer 4: I meant: do you have to talk about `CAMLP5 IDENT` here?

Reviewer: Not necessarily, but the point of the comment is that this shares the same grammar as Coq idents. Should I just write "an identifier"?

- :n:`keyword(s = @string)`:

  + parses the string :n:`s` as a keyword and returns `()`.

- :n:`terminal(s = @string)`:

  + parses the string :n:`s` as a keyword, if it is already a
    keyword, otherwise as an :n:`@ident`. Returns `()`.

- :n:`seq(@scope__1, ..., @scope__2)`:

  + parses :n:`scope__1`, ..., :n:`scope__n` in this order, and produces a tuple made
    out of the parsed values in the same order. As an optimization, all
    subscopes of the form :n:`STRING` are left out of the returned tuple, instead
    of returning a useless unit value. It is forbidden for the various

>     subscopes to refer to the global entry using self of next.

Reviewer 4: ```suggestion
    subscopes to refer to the global entry using self or next.
```

A few other specific scopes exist to handle Ltac1-like syntax, but their use is

> discouraged and they are thus not documented.

Reviewer 4: Doesn't make any sense. If they are internal, they should be documented as such.

Reviewer: Allow me to say the same thing. Documenting internal data is subtly pushing users to rely on it. I know, I'm a specialist of this kind of crime.

For now there is no way to declare new scopes from Ltac2 side, but this is

> planned.

Reviewer 4: Overall, this section is really missing an example.

Reviewer: Isn't that given in the next section? There is a full-fledged example of tactic notation with a quite complex scope.

Notations
~~~~~~~~~

The Ltac2 parser can be extended by syntactic notations.

.. cmd:: Ltac2 Notation {+ @var (@scope) %| @string } {? : @int} := @term
   :name: Ltac2 Notation

   A Ltac2 notation adds a parsing rule to the Ltac2 grammar, which is expanded
   to the provided body where every token from the notation is let-bound to the
   corresponding generated expression.

   For instance, assume we perform:

   :n:`Ltac2 Notation "foo" c(thunk(constr)) ids(list0(ident)) := Bar.f c ids.`

   Then the following expression

>    :n:`let y := @X in foo (nat -> nat) x $y`

Reviewer 4: Once again, this should not be in a `:n:` if `@` is supposed to be part of the Ltac2 syntax.

   will expand at parsing time to

   :n:`let y := @X in`

>    :n:`let c := fun () => constr:(nat -> nat) with ids := [@x; y] in Bar.f c ids`

Reviewer 4: Ditto.

   Beware that the order of evaluation of multiple let-bindings is not specified,
   so that you may have to resort to thunking to ensure that side-effects are
   performed at the right time.

Abbreviations
~~~~~~~~~~~~~

> .. cmdv:: Ltac2 Notation @ident := @term

Reviewer 4: Are you sure this is `ident` and not `lident`. And what about `term`? Should it be `ltac2_term`? Same question for the command defined above and for the command defined below.

  This command introduces a special kind of notations, called abbreviations,
  that is designed so that it does not add any parsing rules. It is similar in
  spirit to Coq abbreviations, insofar as its main purpose is to give an
  absolute name to a piece of pure syntax, which can be transparently referred

>   by this name as if it were a proper definition.

Reviewer 4: I have always doubted of the usefulness of Coq's abbreviation mechanism, and I doubt of the usefulness of this mechanism as well. In particular, this automatic thunking of arguments is very strange.

Reviewer: This is critical to handle macros, you need the argument not to be evaluated to be substituted syntactically.

  The abbreviation can then be manipulated just as a normal Ltac2 definition,
  except that it is expanded at internalization time into the given expression.
  Furthermore, in order to make this kind of construction useful in practice in
  an effectful language such as Ltac2, any syntactic argument to an abbreviation
  is thunked on-the-fly during its expansion.

For instance, suppose that we define the following.

:n:`Ltac2 Notation foo := fun x => x ().`

Then we have the following expansion at internalization time.

:n:`foo 0 ↦ (fun x => x ()) (fun _ => 0)`

Note that abbreviations are not typechecked at all, and may result in typing
errors after expansion.

Evaluation
----------

Ltac2 features a toplevel loop that can be used to evaluate expressions.

.. cmd:: Ltac2 Eval @term
   :name: Ltac2 Eval

   This command evaluates the term in the current proof if there is one, or in the
   global environment otherwise, and displays the resulting value to the user

>    together with its type. This function is pure in the sense that it does not

Reviewer 4: Function?

Reviewer: Wrote "command" instead.

   modify the state of the proof, and in particular all side-effects are discarded.

Debug
-----

.. opt:: Ltac2 Backtrace

>    When this option is set, toplevel failures will be printed with a backtrace.

Reviewer 4: Example?

Compatibility layer with Ltac1
------------------------------

Ltac1 from Ltac2
~~~~~~~~~~~~~~~~

Simple API
++++++++++

One can call Ltac1 code from Ltac2 by using the :n:`ltac1` quotation. It parses
a Ltac1 expression, and semantics of this quotation is the evaluation of the

> corresponding code for its side effects. In particular, in cannot return values,

Reviewer 4: ```suggestion
corresponding code for its side effects. In particular, it cannot return values,
```

or using the `ltac1val` quotation. The latter parses the same as `ltac1`, but
has type `Ltac2.Ltac1.t` instead of `unit`, and dynamically behaves as an Ltac1
thunk, i.e. `ltac1val:(foo)` corresponds to the tactic closure that Ltac1
would generate from `idtac; foo`.

Due to intricate dynamic semantics, understanding when Ltac1 value quotations
focus is very hard. This is why some functions return a continuation-passing
style value, as it can dispatch dynamically between focused and unfocused
behaviour.

Ltac2 from Ltac1
~~~~~~~~~~~~~~~~

Same as above by switching Ltac1 by Ltac2 and using the `ltac2` quotation
instead.

Note that the tactic expression is evaluated eagerly, if one wants to use it as
an argument to a Ltac1 function, she has to resort to the good old
:n:`idtac; ltac2:(foo)` trick. For instance, the code below will fail immediately
and won't print anything.

.. coqtop:: in

   From Ltac2 Require Import Ltac2.
   Set Default Proof Mode "Classic".

.. coqtop:: all

   Ltac mytac tac := idtac "wow"; tac.

   Goal True.
   Proof.
   Fail mytac ltac2:(fail).

Transition from Ltac1
---------------------

> Owing to the use of a bunch of notations, the transition should not be too

Reviewer 4: Bunch is colloquial.

difficult. In particular, it should be possible to do it incrementally. That
said, we do *not* guarantee you it is going to be a blissful walk either.
Hopefully, owing to the fact Ltac2 is typed, the interactive dialogue with Coq
will help you.

We list the major changes and the transition strategies hereafter.

Syntax changes
~~~~~~~~~~~~~~

Due to conflicts, a few syntactic rules have changed.

> - The dispatch tactical :n:`tac; [foo|bar]` is now written :n:`tac > [foo|bar]`.

Reviewer 4: It feels weird to read about this only then.

Reviewer: Where would you put it? Is the doc supposed to be standalone, or should it be understood relatively to Ltac1? We should stick to one or the other.

Reviewer 4: I think it should be standalone mostly. Obviously some parts do refer to Ltac1 for comparison, but it should be possible to skip them. In this case, I think it would make sense to add this construct to the initial ltac2_term grammar.

Reviewer: This is not part of the global grammar. Such a notation is only available when the `Ltac2.Notations` module is imported (which is the case by transitivity if you `Import Ltac2.Ltac2`), but I don't believe this should be part of the default term syntax.

Once again, I believe that part of the documentation should be moved to the `.v` files, where the code that defines the functions and notations lives.

- Levels of a few operators have been revised. Some tacticals now parse as if
  they were a normal function, i.e. one has to put parentheses around the
  argument when it is complex, e.g an abstraction. List of affected tacticals:
  :n:`try`, :n:`repeat`, :n:`do`, :n:`once`, :n:`progress`, :n:`time`, :n:`abstract`.
- :n:`idtac` is no more. Either use :n:`()` if you expect nothing to happen,
  :n:`(fun () => ())` if you want a thunk (see next section), or use printing
  primitives from the :n:`Message` module if you want to display something.

Tactic delay
~~~~~~~~~~~~

Tactics are not magically delayed anymore, neither as functions nor as
arguments. It is your responsibility to thunk them beforehand and apply them
at the call site.

A typical example of a delayed function:

:n:`Ltac foo := blah.`

becomes

:n:`Ltac2 foo () := blah.`

All subsequent calls to `foo` must be applied to perform the same effect as
before.

Likewise, for arguments:

:n:`Ltac bar tac := tac; tac; tac.`

becomes

:n:`Ltac2 bar tac := tac (); tac (); tac ().`

We recommend the use of syntactic notations to ease the transition. For

> instance, the first example can alternatively written as:

Reviewer 4: ```suggestion
instance, the first example can alternatively be written as:
```

:n:`Ltac2 foo0 () := blah.`
:n:`Ltac2 Notation foo := foo0 ().`

This allows to keep the subsequent calls to the tactic as-is, as the
expression `foo` will be implicitly expanded everywhere into `foo0 ()`. Such
a trick also works for arguments, as arguments of syntactic notations are
implicitly thunked. The second example could thus be written as follows.

:n:`Ltac2 bar0 tac := tac (); tac (); tac ().`
:n:`Ltac2 Notation bar := bar0.`

Variable binding
~~~~~~~~~~~~~~~~

Ltac1 relies on complex dynamic trickery to be able to tell apart bound
variables from terms, hypotheses, etc. There is no such thing in Ltac2,
as variables are recognized statically and other constructions do not live in
the same syntactic world. Due to the abuse of quotations, it can sometimes be
complicated to know what a mere identifier represents in a tactic expression. We

> recommend tracking the context and letting the compiler spit typing errors to

Reviewer 4: Spit is colloquial in this context.

Reviewer 3: And a bit gross.

understand what is going on.

We list below the typical changes one has to perform depending on the static
errors produced by the typechecker.

In Ltac expressions
+++++++++++++++++++

> - `Unbound value X`, `Unbound constructor X`:

Reviewer 4: Why don't you use coqrst's `.. exn::` directive?

  * if `X` is meant to be a term from the current stactic environment, replace
    the problematic use by `'X`.
  * if `X` is meant to be a hypothesis from the goal context, replace the
    problematic use by `&X`.

In quotations
+++++++++++++

- `The reference X was not found in the current environment`:

  * if `X` is meant to be a tactic expression bound by a Ltac2 let or function,
    replace the problematic use by `$X`.
  * if `X` is meant to be a hypothesis from the goal context, replace the
    problematic use by `&X`.

Exception catching
~~~~~~~~~~~~~~~~~~

Ltac2 features a proper exception-catching mechanism. For this reason, the

> Ltac1 mechanism relying on `fail` taking integers and tacticals decreasing it

Reviewer 4: ```suggestion
Ltac1 mechanism relying on `fail` taking integers, and tacticals decreasing it,
```

has been removed. Now exceptions are preserved by all tacticals, and it is

> your duty to catch it and reraise it depending on your use.

Reviewer 4: ```suggestion
your duty to catch them and reraise them depending on your use.
```
